Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "URLPatternTokenizer.h"
+
+#include <unicode/uchar.h>
+
+namespace WebCore {
+
+// https://urlpattern.spec.whatwg.org/#is-a-valid-name-code-point
+static bool isValidNameCodepoint(UChar32 codepoint, bool first)
+{
+    if (first)
+        return u_isIDStart(codepoint);
+
+    return u_isIDPart(codepoint);
+}
+
+bool Token::isNull()
+{
+    return (this->type == TokenType::Open && !this->index && this->value.isEmpty());
+}
+
+// https://urlpattern.spec.whatwg.org/#get-the-next-code-point
+void Tokenizer::getNextCodePoint()
+{
+    this->m_codepoint = this->input.characterAt(this->m_nextIndex);
+    this->m_nextIndex++;
```

Comment:
`++m_nextIndex;`

Reply:
But actually, this doesn't look correct. If the spec says to get the next code point then just incrementing an index like this is not correct. This will just get you the next character, not the next code point.

code points are 32 bits and you can take a look at `StringView::CodePoints::Iterator::operator++()` to see how we iterate code points.